src/rp2_common/pico_sleep/sleep.c

```diff
@@ -102,20 +143,77 @@ void sleep_run_from_dormant_source(dormant_source_t dormant_source) {
     setup_default_uart();
 }
 
-// Go to sleep until woken up by the RTC
-void sleep_goto_sleep_until(datetime_t *t, rtc_callback_t callback) {
+static void processor_deep_sleep(void) {
+    // Enable deep sleep at the proc
+#ifdef __riscv
+    uint32_t bits = RVCSR_MSLEEP_POWERDOWN_BITS;
+    if (!get_core_num()) {
+        bits |= RVCSR_MSLEEP_DEEPSLEEP_BITS;
+    }
+    riscv_set_csr(RVCSR_MSLEEP_OFFSET, bits);
+#else
+    scb_hw->scr |= ARM_CPU_PREFIXED(SCR_SLEEPDEEP_BITS);
+#endif
+}
+
+void sleep_goto_aon_sleep_until(struct timespec *ts, aon_timer_alarm_handler_t callback)
+{
+
     // We should have already called the sleep_run_from_dormant_source function
-    assert(dormant_source_valid(_dormant_source));
+    // This is only needed for dormancy although it saves power running from xosc while sleeping
+    //assert(dormant_source_valid(_dormant_source));
 
-    // Turn off all clocks when in sleep mode except for RTC
+#if PICO_RP2040
     clocks_hw->sleep_en0 = CLOCKS_SLEEP_EN0_CLK_RTC_RTC_BITS;
     clocks_hw->sleep_en1 = 0x0;
+#else
+    clocks_hw->sleep_en0 = CLOCKS_SLEEP_EN0_CLK_REF_POWMAN_BITS;
+    clocks_hw->sleep_en1 = 0x0;
+#endif
 
-    rtc_set_alarm(t, callback);
+    aon_timer_enable_alarm(ts, callback, false);
 
-    uint save = scb_hw->scr;
     // Enable deep sleep at the proc
-    scb_hw->scr = save | M0PLUS_SCR_SLEEPDEEP_BITS;
+    processor_deep_sleep();
+
+    // Go to sleep
+    __wfi();
+}
+
+void sleep_goto_sleep_for(uint32_t delay_ms, hardware_alarm_callback_t callback, int *alarm_id)
+{
+    // We should have already called the sleep_run_from_dormant_source function
+    // This is only needed for dormancy although it saves power running from xosc while sleeping
+    //assert(dormant_source_valid(_dormant_source));
+
+    // Turn off all clocks except for the timer
+    clocks_hw->sleep_en0 = 0x0;
+#if PICO_RP2040
+    clocks_hw->sleep_en1 = CLOCKS_SLEEP_EN1_CLK_SYS_TIMER_BITS;
+#elif PICO_RP2350
+    clocks_hw->sleep_en1 = CLOCKS_SLEEP_EN1_CLK_REF_TICKS_BITS | CLOCKS_SLEEP_EN1_CLK_SYS_TIMER0_BITS;
+#else
+#error Unknown processor
+#endif
+
+    int alarm_num = hardware_alarm_claim_unused(true);
+
+    // Return the timer alarm irq number (0..3) so it can be disabled and freed from the main program
+    *alarm_id = alarm_num;
+
+    uart_default_tx_wait_blocking();
```

Comment:
I wonder if it might be worth adding a comment explaining why this line is needed?

Reply:
I think that should be stdio_flush